design/automatic-network-configuration.md

```diff
@@ -0,0 +1,641 @@
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: Switch
+metadata:
+  name: switch-example
+  namespace: default
+spec:
+  provider:
+    kind: AnsibleSwitch
+    name: switch-sample
+  # ports is a map whose key is the name of a SwitchPort CR and value is
+  # some info and limit of the port that SwitchPort CR represents
+  ports:
+    "switchport-example":
+      # The real port name in the switch
+      name: <port-name>
+      # True if this port is not available, false otherwise
+      disabled: false
+      # Indicates the range of VLANs allowed by this port in the switch
+      vlanRange: 1, 6-10
+      # True if this port can be used as a trunk port, false otherwise
+      trunkDisable: false
+```
+
+##### ProviderSwitch CRD
+
+ProviderSwitch CRDs include the login information of switches for different
+protocols. For each ProviderSwitch CRD, there will be a provider
+implementation knowing how to interact with the switch with this information.
+
+###### AnsibleSwitch CRD
+
+Use ansible as the backend to connect to the configuration switch
+(using the network-runner tool, currently only supports the switches supported
+by network-runner).
+
+```yaml
+apiVersion: v1alpha1
+kind: AnsibleSwitch
+metaData:
+  name: switch-sample
+  namespace: default
+spec:
+  # The type of OS this switch runs
+  os: fos
+  # The ip address of the switch
+  ip: 192.168.0.1
+  # The port to use for SSH connection
+  port: 22
+  # Only for ovs switch
+  bridge:
+  # The login credentials of switch
+  credentialsSecret:
+    name: switch-example-secret
+```
+
+##### SwitchPort CRD
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: SwitchPort
+metadata:
+  name: switchport-example
+  ownerReferences:
+    - apiVersion: metal3.io/v1alpha1
+      kind: Switch
+      name: switch-example
+      uid: <switch-uuid>
+spec:
+  configuration:
+    name: switchportconfiguration-example
+    namespect: default
+```
+
+##### SwitchPortConfiguration CRD
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: SwitchPortConfiguration
+metadata:
+  name: switchportconfiguration-example
+spec:
+  # Represents the ACL rules implemented in the switch.
+  acl:
+  - type: # ipv4, ipv6
+    action: # allow, deny
+    protocol: # TCP, UDP, ICMP, ALL
+    src: # xxx.xxx.xxx.xxx/xx
+    srcPortRange: # 22, 22-30
+    des: # xxx.xxx.xxx.xxx/xx
+    desPortRange: # 22, 22-30
+  # Indicates which VLAN this port should be placed in.
+  untaggedVLAN: 1
+  # The range of tagged vlans.
+  vlans: "2-10"
+  disabled: false
+```
+
+##### SwitchResourceLimit CRD
+
+A `SwitchResourceLimit` CR must be created by the administrator for each
+user under their namespace.
+There is only one `SwitchResourceLimit` CR in one namespace, and
+all the `SwitchResourceLimit` CR must use the same name 'user-limit'.
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: SwitchResourceLimit
+metadata:
+  name: user-limit
+  namespace: user1
+spec:
+  # Indicates the range of VLANs allowed by the user
+  vlanRange: 5-10
```

Comment:
@zaneb  We clarify the vlan range that users can use through the vlanRange field. But when the user does not use so many VLANs, other users cannot use it either, which will cause a waste of resources. 
So we think it is better to limit the number of VLANs that users can use, and then combine the mapping [method](https://github.com/metal3-io/metal3-docs/pull/132#discussion_r710014679) we said before to ensure that users will not use the same VLAN. WDYT?